gnomad/utils/sparse_mt.py

```diff
@@ -1025,53 +1050,92 @@ def join_with_ref(mt: hl.MatrixTable) -> hl.MatrixTable:
+                    total_DP=hl.agg.sum(mt.DP),
+                ),
+            ),
+            mt.group_membership,
+        )
+    ).rows()
+    ht = ht.checkpoint(hl.utils.new_temp_file("coverage_stats", "ht"))
 
-    # This expression aggregates the DP counter in reverse order of the coverage_over_x_bins
-    # and computes the cumulative sum over them.
+    # This expression aggregates the DP counter in reverse order of the
+    # coverage_over_x_bins and computes the cumulative sum over them.
     # It needs to be in reverse order because we want the sum over samples
     # covered by > X.
-    count_array_expr = hl.cumulative_sum(
-        hl.array(
-            [
-                hl.int32(coverage_counter_expr.get(max_coverage_bin, 0))
-            ]  # The coverage was already floored to the max_coverage_bin, so no more aggregation is needed for the max bin
-            # For each of the other bins, coverage needs to be summed between the
-            # boundaries
-        ).extend(
-            hl.range(hl.len(hl_coverage_over_x_bins) - 1, 0, step=-1).map(
-                lambda i: hl.sum(
-                    hl.range(
-                        hl_coverage_over_x_bins[i - 1], hl_coverage_over_x_bins[i]
-                    ).map(lambda j: hl.int32(coverage_counter_expr.get(j, 0)))
+    count_array_expr = ht.coverage_stats.map(
+        lambda x: hl.cumulative_sum(
+            hl.array(
+                # The coverage was already floored to the max_coverage_bin, so no more
+                # aggregation is needed for the max bin.
+                [hl.int32(x.coverage_counter.get(max_coverage_bin, 0))]
+                # For each of the other bins, coverage needs to be summed between the
+                # boundaries.
+            ).extend(
+                hl.range(hl.len(hl_coverage_over_x_bins) - 1, 0, step=-1).map(
+                    lambda i: hl.sum(
+                        hl.range(
+                            hl_coverage_over_x_bins[i - 1], hl_coverage_over_x_bins[i]
+                        ).map(lambda j: hl.int32(x.coverage_counter.get(j, 0)))
+                    )
                 )
             )
         )
     )
-    mean_expr = hl.agg.mean(hl.or_else(mt.DP, 0))
-
-    # Annotate rows now
-    ht = mt.select_rows(
-        mean=hl.if_else(hl.is_nan(mean_expr), 0, mean_expr),
-        median_approx=hl.or_else(hl.agg.approx_median(hl.or_else(mt.DP, 0)), 0),
-        total_DP=hl.agg.sum(mt.DP),
-        **{
-            f"over_{x}": count_array_expr[i] / n_samples
-            for i, x in zip(
-                range(
-                    len(coverage_over_x_bins) - 1, -1, -1
-                ),  # Reverse the bin index as count_array_expr has the reverse order
-                coverage_over_x_bins,
-            )
-        },
-    ).rows()
 
+    ht = ht.annotate(
+        coverage_stats=hl.map(
+            lambda c, g: c.annotate(
+                **{
+                    f"over_{x}": g[i] / n_samples
```

Comment:
I might be missing something but wouldnt you want `n_samples` to be specific to the strata instead of using the complete sample count?